accessor/reduced_row_major.hpp

```diff
@@ -222,6 +222,26 @@ class reduced_row_major {
                    stride_};
     }
 
+
+    /**
+     * Returns the storage address for the given indices. If the storage is
+     * const, a const address is returned, otherwise, an address is returned.
+     *
+     * @param indices  indices which value is supposed to access
+     *
+     * @returns  the const address if the accessor is const (if the storage type
+     *           is const), or an address if the accessor is non-const
+     */
+    template <typename... Indices>
+    constexpr GKO_ACC_ATTRIBUTES std::enable_if_t<
+        are_all_integral<Indices...>::value,
+        std::conditional_t<is_const, const storage_type * GKO_ACC_RESTRICT,
+                           storage_type * GKO_ACC_RESTRICT>>
+    get_storage_address(Indices&&... indices) const
+    {
+        return storage_ + compute_index(std::forward<Indices>(indices)...);
+    }
```

Comment:
Could we put the `compute_index`  into the public interface? It's currently in the protected interface. I can simply compute the storage address out of the accessor without touching atomic add.
Will POSIT have its own math operator? (like half) or it can only be used as accessor storage.
if it is like half and we only use the specific length like 32bit, 64 bits, atomic_add still works for it by passing through atomic_cas. 
If it is the storage type of accessor, it will not work with the current atomic_add. Need to use some overload to call atomicCAS

Reply:
The POSIT is simply a storage format, so it can only be converted to and from IEEE types and all arithmetic operations happen in IEEE.
Making `compute_index` public would be fine with me, however, relying on that inside the CSR kernel defeats the purpose of having an Accessor interface since it **only** works with `reduced_row_major`.
That is the reason why I would prefer to have another function that handles support for more than one accessor. Inside that function, it would be fine to rely on `get_storage_address` or `compute_index`.

I need to redesign the whole accessor structure to make this more clear and easier to use, which I will do after finishing the block-compressor accessor.

Reply:
Not as a `ValueType`, but as a storage format. You get the same problem when you want to add `bfloat16` and other storage formats to it, which I would like to have in the future.

Reply:
For our matrices, `ValueType` *is* the storage type, and changing that would require significant modifications of Ginkgo's classes, so I'm not sure this is warranted (yet).

Reply:
For the CSR matrix, you are correct, but I don't want to have 2 implementations that pretty much to the same thing.
Currently, I see 2 options to tackle this and similar problems (that need to be implemented):
1. Request that the accessor has either built-in atomic support or means to give a pointer to the underlying storage format
2. Overload the `atomic_add` to work with the accessor
Of course, you could also remove the accessor all together and implement the `storage_type` and `arithmetic_type` templates by hand, but that doesn't scale well.

Reply:
@yhmtsai For now, I think we can also leave it as you implemented it and I will change it in 1-2 months (when I update the accessor interface).

Reply:
I see. It's not the public interface, so we can change it in any time.
For me, the second option will be like `atomic_add(accessor, value, index...);` then unpack the information there

Reply:
@thoasm Can you elaborate how we would have scaling issues? Assuming we take the storage types from the input LinOps and use the largest common type for arithmetic_type, and all types are interconvertible, it looks to me like we could use the same implementation for all instantiations.
As Mike mentioned, `atomic_add` only makes sense performance-wise where we have hardware support, and those are also the cases where the operations boil down to a `static_cast` when accessing the input/output vectors.

Reply:
Not sure what scaling issues you mean @upsj . The 2 implementations I was referring to was the accessor and the CSR SpMV.

Reply:
I meant the comment "but that doesn't scale well"

Reply:
Oh, that "scaling" was referring to having multiple implementation that do the same thing, which leads to bad maintainability.
I didn't mean performance scaling, sorry for the confusion.

Reply:
Since this is likely not the only place where we need some atomic access, I would prefer of solving it with the accessor once and use that in more places.
But if you think it is better to implement it without the accessor, feel free to do so.